File: parking spot booking/backend/src/main/java/com/parkingfinder/backend/service/BookingService.java
package com.parkingfinder.backend.service;

import com.parkingfinder.backend.dto.booking.BookingHistoryResponse;
import com.parkingfinder.backend.dto.booking.BookingResponse;
import com.parkingfinder.backend.dto.booking.CheckoutRequest;
import com.parkingfinder.backend.dto.booking.ConfirmBookingRequest;
import com.parkingfinder.backend.dto.booking.CreateReservationHoldRequest;
import com.parkingfinder.backend.dto.booking.ReservationHoldResponse;
import java.util.List;

/**
 * Booking lifecycle management: reservation holds, confirmation, cancellation,
 * checkout, and audit history.
 */
public interface BookingService {

    /** Create a short-lived reservation hold on a slot (10-minute expiry). */
    ReservationHoldResponse createReservationHold(CreateReservationHoldRequest request);

    /** Confirm and pay for a booking (wallet debit + payment record). */
    BookingResponse confirmBooking(ConfirmBookingRequest request);

    /** List all bookings for the currently-authenticated driver. */
    List<BookingResponse> getMyBookings();

    /** Get a single booking by ID (must belong to current user). */
    BookingResponse getBookingById(String bookingId);

    /** Cancel an upcoming or active booking and refund the wallet. */
    BookingResponse cancelBooking(String bookingId);

    /** Check out from a booking; calculates late fees if applicable. */
    BookingResponse checkoutBooking(String bookingId, CheckoutRequest request);

    /** Full status-change audit trail for a booking. */
    List<BookingHistoryResponse> getBookingHistory(String bookingId);
}

File: parking spot booking (final project)/frontend/src/services/bookingService.js
import apiClient from "./apiClient";
import { unwrap, toLocalDateTime } from "./apiHelpers";

export const bookingService = {
  async getBookingsByUser() {
    const res = await apiClient.get("/bookings/me");
    return unwrap(res) || [];
  },

  async getBookingById(bookingId) {
    const res = await apiClient.get(`/bookings/${bookingId}`);
    return unwrap(res);
  },

  async getBookingHistory(bookingId) {
    const res = await apiClient.get(`/bookings/${bookingId}/history`);
    return unwrap(res) || [];
  },

  async createReservationHold(payload) {
    const res = await apiClient.post("/bookings/holds", payload);
    return unwrap(res);
  },

  async confirmBooking(payload) {
    const body = {
      ...payload,
      booked_start_time: toLocalDateTime(payload.booked_start_time),
      booked_end_time: toLocalDateTime(payload.booked_end_time),
    };
    const res = await apiClient.post("/bookings/confirm", body);
    return unwrap(res);
  },

  async cancelBooking(bookingId) {
    const res = await apiClient.post(`/bookings/${bookingId}/cancel`);
    return unwrap(res);
  },

  async checkoutBooking(bookingId, actualCheckoutTime = new Date()) {
    const res = await apiClient.post(`/bookings/${bookingId}/checkout`, {
      actual_checkout_time: toLocalDateTime(actualCheckoutTime),
    });
    return unwrap(res);
  },

  async submitReview(payload) {
    const res = await apiClient.post("/reviews", payload);
    return unwrap(res);
  },

  async getReviewsByUser() {
    const res = await apiClient.get("/reviews/me");
    return unwrap(res) || [];
  },
};
